api/files/serializers.py

```diff
@@ -188,6 +213,23 @@ def user_id(self, obj):
 
     def update(self, instance, validated_data):
         assert isinstance(instance, FileNode), 'Instance must be a FileNode'
+        if instance.provider == 'osfstorage':
+            auth = get_user_auth(self.context['request'])
+            old_tags = set([tag._id for tag in instance.tags])
+            if 'tags' in validated_data:
+                current_tags = set(validated_data.pop('tags', []))
+            elif self.partial:
+                current_tags = set(old_tags)
+            else:
+                current_tags = set()
+
+            for new_tag in (current_tags - old_tags):
+                instance.add_tag(new_tag, auth=auth)
+            for deleted_tag in (old_tags - current_tags):
+                instance.remove_tag(deleted_tag, auth=auth)
+        if instance.provider != 'osfstorage' and 'tags' in validated_data:
```

Comment:
I think if you bump this to the top of the method just below the assert, then you won't need the lines above this to be in an if branch.